### PR TITLE
Permit different foreground tables for different instances of hierarchical removal.

### PR DESCRIPTION
**Pull request by stevereyes01**
Update the foreground table so that it can display quantities for hierarchical removals and still have the same parameters sorted correctly. 

This requires a bit of sorting manipulation between the information given in the single detector trigger information, the bank information, and the statmap coinc file information.

Once this is merged, I'll need to fix `plotting.py` and `make_coinc_search_workflow` to accept options to results pages for hierarchical removals.

**A**
Table without hierarchical removal on all of O1:
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-original.html

This runs the code using a statmap file that did not have any hierarchical removals permitted.

**B**
Table with all hierarchical removals done which were possible but final results displayed (default):
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-default.html

If no options are given to the foreground table code, the table generated gives the ifars as regenerated by running `pycbc_coinc_statmap` with some maximum number of hierarchical removals. If no hierarchical removals could or were done, this is default the same foreground table that we would expect to see.

**C**
Table before any hierarchical removals were done. This implies that later hierarchical removals were done, otherwise this table cannot be generated:
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h0.html

**D**
Table after 1 hierarchical removal was done:
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h1.html

**E**
Table after 2 hierarchical removals were done:
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h2.html

No further hierarchical removals could be done, so this is the most we can do.

The code is run with the following options and help message:

```
usage: pycbc_page_foreground [-h] [--version] [--trigger-file TRIGGER_FILE]
                             [--bank-file BANK_FILE]
                             [--single-detector-triggers SINGLE_DETECTOR_TRIGGERS [SINGLE_DETECTOR_TRIGGERS ...]]
                             [--verbose] [--output-file OUTPUT_FILE]
                             [--foreground-tag FOREGROUND_TAG]
                             [--num-coincs-to-write NUM_COINCS_TO_WRITE]
                             [--h-num-iterations H_NUM_ITERATIONS]

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  --trigger-file TRIGGER_FILE
  --bank-file BANK_FILE
  --single-detector-triggers SINGLE_DETECTOR_TRIGGERS [SINGLE_DETECTOR_TRIGGERS ...]
  --verbose
  --output-file OUTPUT_FILE
  --foreground-tag FOREGROUND_TAG
  --num-coincs-to-write NUM_COINCS_TO_WRITE
  --h-num-iterations H_NUM_ITERATIONS
                        Indicate which FARs and FAPs (FANs) to write based on
                        the number of hierarchical removals done. Choosing
                        None defaults to giving the FANs after all
                        hierarchical removals were done depending on previous
                        configuration choices. Choosing 0 selects writing the
                        FANs prior to any hierarchical removals. Choosing 1
                        means writing the FANs after doing 1 hierarchical
                        removal. The program will fail if the user selects a
                        number above the number of hierarchical removals done.
                        [default=None]
```

### CODE REVIEW

**Comment by stevereyes01** [comment]
The proper way to do this is to let `ForegroundTriggers` in `hdf.py` in the `pycbc.io` also have a `group='Foreground_h%' % h_iteration_num` in https://github.com/ligo-cbc/pycbc/blob/master/pycbc/io/hdf.py#L576-L593

but this would require saving more data to 'foreground_hX' in `pycbc_coinc_statmap` and also requires some skill in editing that class that I don't really understand very well.

**Comment by stevereyes01** [comment]
There's no bug in the timing differences. I just don't know how to read the tables properly. GW150914 and GW151226 switch places in some of these tables, but they all consistently change places in the way that you would want. GW150914 has a 7.32 ms difference in arrival times between H1 and L1, but GW151226 and LVT151012 both have 0.98 ms difference in arrival times between H1 and L1.

**Comment by stevereyes01** [comment]
I'll take your questioning as desire to review @spxiwh

**Comment by spxiwh** [comment]
@stevereyes01 Thanks for putting this together, but I think I'd like to see some of the complexity here moved away into the hdf module so it could be used elsewhere. Am I right that the point here, if the new option is given, is to make the result table with the loudest N triggers not present and significances quoted accordingly?

I think I would suggest making a method of ForegroundTriggers called something like:
```
ignore_n_loudest_triggers
```
which would update the mask attribute of the class to not read the n loudest triggers. Then the code in the executable would just be:
```
fortrigs = hdf.ForegroundTriggers(args.trigger_file, args.bank_file,
                                  sngl_files=args.single_detector_triggers,
                                  n_loudest=args.num_coincs_to_write)
if args.h_num_iterations:
    fortrigs.ignore_n_loudest_triggers(args.h_num_iterations)
```
and then the only other change needed in the executable would be a switch on which `stat` variable to use.

Some other things:

  * Please update the help message for h-num-iterations not to mention "FAN". P-value is the correct term, and IMO it *should* be used (even if it means we have both terms used within PyCBC).
  * Please split off the change to rename the results table to p-value to a separate patch. (If FAP appears anywhere else in the results page, please change there as well).
  * h_num_iterations/`--h-num-iterations` as an option and a variable name is not clear to me. Can you name this something clearer?

**Comment by stevereyes01** [comment]
Thanks Ian, these are really helpful comments. I'll make these edits.

**Comment by stevereyes01** [comment]
Oops, and yes, to answer your question, that is the purpose of this code: "make the result table with the loudest N triggers not present and significances quoted accordingly"

**Comment by stevereyes01** [comment]
@spxiwh PR is updated. Just note that when doing hierarchical removal the default ordering for the table becomes sorting by ranking statistic rather than by ifar which is the default when not running with hierarchical removal. I don't really think it matters to be honest. The numbers are all sorted properly. Let me know if the code looks appropriate now.

File: bin/hdfcoinc/pycbc_page_foreground [hunk 1]
```diff
@@ -31,32 +59,62 @@ if args.output_file.endswith('.xml') or args.output_file.endswith('.xml.gz'):
...
     time2 = fortrigs.get_coincfile_array('time2')
-
     mass1 = fortrigs.get_bankfile_array('mass1')
     mass2 = fortrigs.get_bankfile_array('mass2')
     spin1z = fortrigs.get_bankfile_array('spin1z')
     spin2z = fortrigs.get_bankfile_array('spin2z')
 
+    # Write the correct false alarm rates and other significance values
+    # to the table after h_num_rm hierarchical removals
+    if h_num_rm >= 0:
```
Comment (ahnitz) [inline bin/hdfcoinc/pycbc_page_foreground]: Steven, this seems dangerous. Why aren't you simply using the trigger_id columns that are written out? This is what the class underyling is doing. As Ian says, the best thing would be to move things to the underlying class. You an add another argument that tells it what key group to use, etc. 

Just save out the same information from pycbc_coinc_statmap like you did for the background groups. https://github.com/ligo-cbc/pycbc/blob/master/bin/hdfcoinc/pycbc_coinc_statmap#L308

File: pycbc/io/hdf.py [hunk 2]
```diff
@@ -648,6 +648,15 @@ def get_snglfile_array_dict(self, variable):
             return_dict[ifo] = curr
         return return_dict
 
+    def ignore_x_loudest(self, num_ignore):
```
Comment (ahnitz) [inline pycbc/io/hdf.py]: If you do the former change, correct me if I'm wrong, but you don't need this function anymore.

File: bin/hdfcoinc/pycbc_page_foreground (hunk 1 above)
Comment (stevereyes01) [inline bin/hdfcoinc/pycbc_page_foreground]: Hey Alex, thanks for commenting, yeah, this is the way I should have done this, you're right. I'll fix this.

File: pycbc/io/hdf.py (hunk 2 above)
Comment (stevereyes01) [inline pycbc/io/hdf.py]: That would be correct.

**Comment by spxiwh** [comment]
I think this is waiting for changes in response to Alex's comments. I'll wait until that is resolved before looking over this again.

**Comment by stevereyes01** [comment]
@ahnitz Alright, now the code uses one of the foreground trigger ids to make the swaps of values for the ifar and fap values. Sorry this took a while to do, I was playing around with trying to fix the stat map code and foreground id's but none of the changes I made seemed to really pan out. So this is just the simplest way to enact more or less the exact same behavior.

I still use the `ignore_x_loudest(self, num_ignore)` function since I think it's still helpful, but I can remove it and just use the foreground trigger id's like you said before.

This PR, now depends on https://github.com/ligo-cbc/pycbc/pull/1653.

@spxiwh All of the tables have now been updated. None of the results have changed, the implementation is just a bit different.

**Comment by spxiwh** [comment]
Is this PR now ready for consideration again? @ahnitz are you happy with this now? Because I've gotten confused with this a little bit, can I ask that this be briefly discussed on Wednesday's call, with the idea being to get this merged then if everyone is happy.

**Comment by stevereyes01** [comment]
@spxiwh I'll leave this to Alex with whether he's happy with it. If you have any questions about the code, feel free to just air your questions here.

File: pycbc/io/hdf.py (hunk 2 above)
Comment (ahnitz) [inline pycbc/io/hdf.py]: Again, I don't believe this should be needed.

File: bin/hdfcoinc/pycbc_page_foreground [hunk 3]
```diff
@@ -31,32 +59,63 @@ if args.output_file.endswith('.xml') or args.output_file.endswith('.xml.gz'):
...
+        # trigger id's from fortrigs.get_coincfile_array and stat file
+        trig_id1_table = fortrigs.get_coincfile_array('trigger_id1')
+
+        trig_id1_hrm = f['foreground_h%s/trigger_id1' % h_num_rm][:]
+        trig_id2_hrm = f['foreground_h%s/trigger_id2' % h_num_rm][:]
+
+        ifar_hrm = f['foreground_h%s/ifar' % h_num_rm][:]
+        fap_hrm = f['foreground_h%s/fap' % h_num_rm][:]
+
+        idx1_list = []
```
Comment (ahnitz) [inline bin/hdfcoinc/pycbc_page_foreground]: Why do we need the next two sections? We've read in the ifar and fap above, and have the trigger id values to look up the rest? Shouldn't this be upfront, and combined with the previous section so we also don't need the call to remove some arbitrary number.

Comment (stevereyes01) [inline bin/hdfcoinc/pycbc_page_foreground]: What I'm doing is read in the ifar and fap, and the trigger ids from `f['foreground_h%s/trigger_id1' % h_num_rm]`. Then I match the trigger ids from this layer of the hdf to the ones at the top level `f['foreground/trigger_id1']`. Using the map between the locations of these indices in each list, replace the little-dogs-in ifar and fap according to that mapping.

The call to remove x loudest removes all of the trigger information associated with things that shouldn't be displayed at the X-th stage of hierarchical removal. I'm not exactly sure what sort of implementation you have in mind, maybe I'm missing something obvious here.

Comment (ahnitz) [inline bin/hdfcoinc/pycbc_page_foreground]: @stevereyes01 What I'm saying is *do not* match the ids, just use them directly. And at the top of the code simply select which trigger ids to use depending on if you are making the hierarchically removed table or not.

Comment (ahnitz) [inline bin/hdfcoinc/pycbc_page_foreground]: The foreground_h%s columns should not have the removed events, so you don't need to have an extra function to remove then.

**Comment by spxiwh** [comment]
@ahnitz @stevereyes01 Will you be working on this while both in the same place? It would be good to get this closed out?

**Comment by spxiwh** [comment]
Any updates on this? If there's nothing new the next time I look over old pull requests then I'll close this ... It's easy to open a new pull request later.

**Comment by stevereyes01** [comment]
Getting back on this, sorry for dropping this.

**Comment by stevereyes01** [comment]
The latest version I just pushed now reproduces the results at the beginning of this thread:

Here are the links:

**A: Running the script in it's default stage (with all possible hierarchical removals done)**
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/third_attempt/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-default.html

**B: Running the script with the option to look at results prior to all hierarchical removals**
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/third_attempt/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h0.html

**C: Running the script with the option to look at results after only 1 hierarchical removal**
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/third_attempt/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h1.html

**D: Running the script with the option after 2 hierarchical removals were done**
https://www.atlas.aei.uni-hannover.de/~steven.reyes/LSC/O2/test_html_h_removal/third_attempt/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h2.html

Hopefully this is a lot closer to a final version of this code. I know that the `pycbc_coinc_statmap` code could be cleaned up, but I think using this implementation makes `pycbc_page_foreground` easier to read. There's a few permutations that one could use here to make this code run, but hopefully this is alright to submit.

**Comment by stevereyes01** [comment]
How to reproduce the html pages produced here:

file locations on atlas to gsiscp:
statmap file: `/home/steven.reyes/projects/cbc/stat-map-test/html_test/clustered-attempt2.hdf`

single detector triggers, 2 files: `/home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/H1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/L1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf`

bank file: `/home/steven.reyes/projects/cbc/stat-map-test/outputpt/bank/H1L1-BANK2HDF-1126051217-11203200.hdf`

You'll have to change the executable path to your virtualenv path, and the output location too.

```
python /home/steven.reyes/pycbc-dev-new-ATLAS/src/pycbc/bin/hdfcoinc/pycbc_page_foreground \
--bank-file /home/steven.reyes/projects/cbc/stat-map-test/outputpt/bank/H1L1-BANK2HDF-1126051217-11203200.hdf  \
--foreground-tag full_data \
--trigger-file clustered-attempt2.hdf \
--single-detector-triggers /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/H1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/L1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf \
--output-file /home/steven.reyes/WWW/LSC/O2/test_html_h_removal/third_attempt/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-default.html \
--num-coincs-to-write 25 \
--verbose
echo "Done1"

python /home/steven.reyes/pycbc-dev-new-ATLAS/src/pycbc/bin/hdfcoinc/pycbc_page_foreground \
--bank-file /home/steven.reyes/projects/cbc/stat-map-test/outputpt/bank/H1L1-BANK2HDF-1126051217-11203200.hdf  \
--foreground-tag full_data \
--trigger-file clustered-attempt2.hdf \
--single-detector-triggers /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/H1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/L1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf \
--hierarchical-rm-iterations 0 \
--output-file /home/steven.reyes/WWW/LSC/O2/test_html_h_removal/third_attempt/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h0.html \
--num-coincs-to-write 25 \
--verbose 
echo "Done2"

python /home/steven.reyes/pycbc-dev-new-ATLAS/src/pycbc/bin/hdfcoinc/pycbc_page_foreground \
--bank-file /home/steven.reyes/projects/cbc/stat-map-test/outputpt/bank/H1L1-BANK2HDF-1126051217-11203200.hdf  \
--foreground-tag full_data \
--trigger-file clustered-attempt2.hdf \
--single-detector-triggers /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/H1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/L1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf \
--hierarchical-rm-iterations 1 \
--output-file /home/steven.reyes/WWW/LSC/O2/test_html_h_removal/third_attempt/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h1.html \
--num-coincs-to-write 25 \
--verbose

echo "Done3"

python /home/steven.reyes/pycbc-dev-new-ATLAS/src/pycbc/bin/hdfcoinc/pycbc_page_foreground \
--bank-file /home/steven.reyes/projects/cbc/stat-map-test/outputpt/bank/H1L1-BANK2HDF-1126051217-11203200.hdf  \
--foreground-tag full_data \
--trigger-file clustered-attempt2.hdf \
--single-detector-triggers /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/H1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf /home/steven.reyes/projects/cbc/stat-map-test/outputpt/full_data/L1-HDF_TRIGGER_MERGE_FULL_DATA-1126051217-11203200.hdf \
--hierarchical-rm-iterations 2 \
--output-file /home/steven.reyes/WWW/LSC/O2/test_html_h_removal/third_attempt/H1L1-PAGE_FOREGROUND_HTML-1126051216-11203201-h2.html \
--num-coincs-to-write 25 \
--verbose

echo "Done4"
```

**Pull request merged**